codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/AuthenticationSchemeGenerator.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import software.amazon.smithy.model.shapes.ServiceShape;
+
+/**
+ * Used by integrations to generate authentication
+ * scheme resolution.
+ *
+ */
+public interface AuthenticationSchemeGenerator {
```

Comment:
wanted to get folks' thoughts on this. creating an interface and making a getter on `GoIntegration` rather than embedding `renderEndpointBasedAuthSchemeResolution` directly into `GoIntegration`. im figuring this class can contain future auth scheme resolution changes?

maybe this is likely overkill for now since we dont know how SRA auth stuff will look in Go yet, so might as well just embed `renderEndpointBasedAuthSchemeResolution` into `GoIntegration`?

Reply:
im guess that the intention is that the `EndpointPropertyRenderer` look something like this:

```java
public interface EndpointPropertyRenderer {
    default void renderEndpointProperty(GoWriter writer, Model model) {
        // pass
    }
}
```
i think this could potentially be a problem with forcing every future property into a specific signature. like what if a new property needs an additional parameter? i suppose this is fine while smithy-go isnt GA. but if we were, we would have to add a new overloaded method to the `EndpointPropertyRenderer`. would this be cleaner than adding a new interface for the property with its own GoIntegration overridable method? thoughts?

Reply:
discussed offline. there was a misunderstanding on how this integration was used. we decided to go with `renderPostEndpointResolutionHook` on the GoIntegration. its what kotlin did and its more amenable to us removing the endpoint-based auth scheme resolution stuff later with planned work for this year.